keps/sig-windows/20190424-windows-cri-containerd.md

```diff
@@ -171,14 +177,236 @@ With virtual-based allocations and Hyper-V, it should be possible to increase th
 
 The work needed will span multiple repos, SIG-Windows will be maintaining a [Windows CRI-Containerd Project Board] to track everything in one place.
 
-#### Proposal: Use Runtimeclass Scheduler to simplify deployments based on OS version requirements
 
-As of version 1.14, RuntimeClass is not considered by the Kubernetes scheduler. There’s no guarantee that a node can start a pod, and it could fail until it’s scheduled on an appropriate node. Additional node labels and nodeSelectors are required to avoid this problem. [RuntimeClass Scheduling](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/runtime-class-scheduling.md) proposes being able to add nodeSelectors automatically when using a RuntimeClass, simplifying the deployment.
+#### Options for supporting multiple OS versions
+
+
+
+
+##### Option A - Use RuntimeClass per Windows OS version
+
+With process isolation, Windows containers must be constrained to only run on a matching Windows Server version node.
+
+[RuntimeClass Scheduling] has a few useful features that can make it easier to match a Windows container to a compatible Windows node. Here's an example of steps that could use RuntimeClass to avoid os version mismatches: 
+
+1. A RuntimeClass such as `WindowsServer1809` defined with [RuntimeClass Scheduling] `nodeSelector` for `kubernetes.io/os=windows` and `nodeSelector` for `beta.kubernetes.io/osversion=Windows1809` set.
+1. Nodes must be labeled with the `kubernetes.io/os=windows` (already done), as well as the additional label `beta.kubernetes.io/osversion=Windows1809`
+
+This would need to be repeated for each Windows version added to the cluster such as `WindowsServer1809`, `WindowsServer1903` and future releases. The cluster operator would need to ensure that nodes are added and scaled as needed for each Windows version independently.
+
+A user doing a deployment must set an appropriate `runtimeClassName` for their Pod to be scheduled on a compatible Windows node.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: mypod
+spec:
+  runtimeClassName: WindowsServer1809
```

Comment:
An advantage of this approach is that by making the runtimeClassName explicit, the user must update the deployment which:
- Forces them to ensure they have an image built for the new versions
- Makes it clear what version they're expecting, and which version a given pod is running
- Automatically triggers a rolling update to move to the new version, with the supported workflows

This disadvantage is it places more work on the dev teams to keep their deployments up-to-date, and also has a risk of a combinatorial explosion of runtime classes.

Reply:
Interesting point. It also enables us to have different ideas around things like `RuntimeOverhead` and such per version. 

@tallclair - Given your feedback would you recommend that we do not have a handler that is "versionless"? IE: no `runhcs-wcow-process` but only `runhcs-wcow-process-1809` etc